Added PowerLauncher spec and images

jyuwono opened the pull request:
## Summary of the Pull Request
Adding PowerLauncher spec and images under doc/specs/images/PowerLauncher/.

this is for Issue #44 

**We'll be closing feedback here 1/29/2020**

crutkas commented:
To view visually, currently, go to  https://github.com/jyuwono/PowerToys/blob/PowerLauncher/doc/specs/44-PowerLauncher.md

enricogior commented:
Just making sure all new contributors are aware of [Github Workflow](https://github.com/microsoft/PowerToys/tree/master/doc/devdocs#github-workflow) indication:
> Use the `Squash and merge` option to merge a PR, if you don't want to squash it because there are logically different commits, use `Rebase and merge`.

Thank you.

jyuwono commented on doc/specs/44-PowerLauncher.md:
```diff
@@ -0,0 +1,480 @@
+# PowerLauncher  
```
Done. The link to view visually is now https://github.com/jyuwono/PowerToys/blob/PowerLauncher/doc/specs/QuickLauncher.md

jyuwono commented:
The link to view visually is now https://github.com/jyuwono/PowerToys/blob/PowerLauncher/doc/specs/QuickLauncher.md

Kaedin commented:
How about checking and including 'ueli' (https://github.com/oliverschwendener/ueli) in the comparison matrix as well? 😊

ghosttie commented:
Another app to compare (from back in the day 👴) [AppRocket](https://www.candylabs.com/approcket) 🚀

crutkas commented:
@kaedin, does ueli provide something other than maybe xplat that our matrix currently doesn't?  Looked very close to Wox.  A comment somewhere made me think it runs electron under the hood

LoZeno commented on doc/specs/QuickLauncher.md:
```diff
@@ -0,0 +1,398 @@
+|[Alfred (free)](https://www.alfredapp.com/)       | ![alt text][cross] | ![alt text][check] |![alt text][cross] | ![alt text][cross] | ![alt text][check]                  | ![alt text][cross] |
+|[Spotlight](https://support.apple.com/en-us/HT204014)</p>       | ![alt text][check] | ![alt text][check] |![alt text][cross] | ![alt text][cross] | ![alt text][check]                  | ![alt text][check] |
+|[Listary](https://www.listary.com/)</p>         | ![alt text][cross] | ![alt text][check] |![alt text][cross] | ![alt text][check] shown as a separate cmd | ![alt text][cross]                  | ![alt text][check] on right click |
+|[Wox](http://www.wox.one/)</p>             | ![alt text][cross] | ![alt text][check] |![alt text][check] via “>” or Win+R | ![alt text][check] on right click | ![alt text][check]                  | ![alt text][check] on right click|
```
Can I point out that on Wox you can access the "run as administrator" and "open file location" with Shift+Enter, not just right click? The whole point of a launcher like this is to be able to keep both hands on the keyboard

jyuwono replied:
Thanks for pointing this out! I've added it to the comparison table.

fdncred commented:
Oh, please don't create this with an indexer. That's all I need on my system is another indexer. How about something that queries the MFT like voidtools' [Everything](https://www.voidtools.com/). There's even an open source [clone ](https://github.com/yuzhengyang/Everything/)of this technique.

PastorGL commented:
Hey @jyuwono,

Win+Space hotkey by default is a bad choice for international users as it is already assigned to the input language switcher. It's better to choose something that doesn't clash.

jyuwono commented:
> Hey @jyuwono,
> 
> Win+Space hotkey by default is a bad choice for international users as it is already assigned to the input language switcher. It's better to choose something that doesn't clash.

Good call out, thank you! I've changed the hotkey default to Alt+Space in the spec, and it can still be customized through the settings page.

JosephTico commented:
I would also suggest to take some lessons from the beta version of Listary.

Some features I really like is that you can quick select a result by using a Ctrl+Number shortcut:

![image](https://i.imgur.com/nEZPEUh.png)

Also having to press Shitf+F10 to open the context menu doesn't seem to intuitive for me, a better solution would be to open it by just pressing the Right arrow key. Also Listary (beta) does the context incredibly well as it loads all the options that selecting a file in Explorer would have:

![image}(https://i.imgur.com/xZ1UrqE.png)
![image](https://user-images.githubusercontent.com/8814190/73126004-0079e980-3f73-11ea-8962-795a85a40569.png)

crutkas commented:
> > Hey @jyuwono,
> > Win+Space hotkey by default is a bad choice for international users as it is already assigned to the input language switcher. It's better to choose something that doesn't clash.
> 
> Good call out, thank you @PastorGL! I've changed the hotkey default to Alt+Space in the spec, and it can still be customized through the settings page.

Needs to be a winkey. We should call this out / make it configurable. Let’s see maybe if we can detect the winkey+space conflict.

PastorGL commented:
@jyuwono 
Alt+Space is equally bad because it is a current-window system menu shortcut since early Windows versions.

@crutkas
There is a list at https://support.microsoft.com/en-us/help/12445/windows-keyboard-shortcuts — looks like almost all Winkey shortcuts are assigned already. Win+\ (backward slash) is the only unused chord on the main keyboard area.

LoZeno commented:
> Also having to press Shitf+F10 to open the context menu doesn't seem to intuitive for me, a better solution would be to open it by just pressing the Right arrow key.

Shift+F10 is consistent with the rest of the OS though, as it's the key combination to open the context menu basically everywhere.

crutkas commented:
> @JosephTico
> 
> > Also having to press Shitf+F10 to open the context menu doesn't seem to intuitive for me, a better solution would be to open it by just pressing the Right arrow key.
> 
> Shift+F10 is consistent with the rest of the OS though, as it's the key combination to open the context menu basically everywhere.

Yup, why we picked Shift-F10.  the keyboard context menu would work as well.

zspitz commented:
When the system has multiple keyboard layouts installed, map the entered text to the keys pressed, and search for the same key presses across keyboards.

For example, when the standard Hebrew keyboard is active, pressing the keys <kbd>c</kbd><kbd>o</kbd><kbd>d</kbd><kbd>e</kbd> results in the entered text `בםגק`. This could be mapped back for the Hebrew keyboard to `code`, and both terms (`בםגק` and `code`) would be searched for.

crutkas commented:
@zspitz this is something we will investigate.  I love the idea, now to see how we pull it off :)

HikaruSora commented:
A few suggestions/thoughts/wants (some of which echo existing feedback) as a WoX user on Windows and former Spotlight user on macOS:

### Search Mode

A search-mode feature, similar to what something like Visual Studio Code uses (whatever they call it), might be worth considering. For context, I'm referring to the ability to search by a specific class of _thing_ much like how VSCode's `Ctrl+P` opens up the palette for file search, `Ctrl+Shift+P` for command search, and `Ctrl+Shift+O` for symbol search. Users can change the mode by starting the query with a particular symbol (for example, `Ctrl+P` opens up the entry field empty but starting the search with `>` changes it to command search).

An example in the scope of PowerLauncher: users looking for files specifically and may not want to be bombarded by apps or commands which have similar names. Similarly, perhaps they're looking only for the results coming from a particular plugin (file search vs image search vs web search vs command search vs ....) and want to reduce noise in the list.

### Copy as Path

A common workflow my colleagues and I do is to open explorer, find the target file, `Shift+Right Click` it, and `Copy as path` for pasting into whatever command we have in an open PowerShell or CMD window. 

Having the corresponding option in PowerLauncher's context menu would pair nicely with the file search functionality: `Fuzzy Find File -> Context Menu -> Copy as path` rather than `Fuzzy Find File -> Context Menu -> Open File Location (?) -> Shift+Right Click -> Copy as Path`

Additionally, having a keyboard shortcut like `Ctrl+Shift+C` to directly copy the focused result's file path to the clipboard would be a plus

### Preview

I saw Niels's Fluent Design mockup of the launcher and it looks fantastic. However, looking at it and comparing to WoX and the other 3rd party launchers, I came to realize that one of the big things  that they tend to lack is some sort of preview pane like macOS's Spotlight or another launcher, [Cerebro](https://github.com/KELiON/cerebro). Having a preview makes it easier to check if a particular result indeed contains the information and provides some additional context that can't fit in a single result entry.

### Result Density

From requirement 13, 

> Users can see the icon of the application/file/folder next to the application/file/folder name

there's no mention about the actual icon size. The mockup and alternate launchers imply that it would be as tall as the result entry and I can't help but think that there should be the option for... the opposite, actually. The nice big results with their icons look really pretty but reduces the number of items that can be displayed at a given time. There's mention of changing text size in Settings requirement 9 but nothing about the icon nor collapsing a result into a single line.

It seems appropriate to give users the option to increase density, perhaps to something around the size of Spotlight's results, to tilt it towards the Function side of the Function vs Form tradeoff. These _are_ _**power**_ users, after all. A dense, small/no-icon, no-nonsense view mode would match some of them better. It may not fit well with the Fluent Design concept but, well, you might be tossing them directly into PowerShell anyways. 😛 

### File Indexing

The document implies that investigation is still ongoing in regards to selecting a file indexer. Perhaps the plugin model could be leveraged to add alternate ones; if a plugin can specify in a json manifest that they are a `fileIndexer` type or something, the user can go to the settings and select it as the primary file indexer over another (to avoid duplicate results).

From what's there so far, it looks like the two possible candidates are the built-in Windows 10 indexer (implied by the Supporting Documents section) and `Everything Search` (weakly implied by the potential WoX collaboration).

From personal usage, I'd strongly suggest considering the latter over the former. The built-in indexer leaves quite a lot to be desired in both performance and indexing breadth. Specifically, through a mini-rant:

* Performance when indexing systems with many millions of files (I've tested as few as 1 million and as many as 10+ million) can take on the order of days, even without indexing file contents. In contrast, Everything Search completed indexing on the same systems in a matter of minutes.
* Indexing is sometimes hit-or-miss with some filetypes/extensions, a common scenario being configuration dotfiles (likely treated as hidden). Sometimes, we want to index everything under the sun, _and_ the sun, and an indexer should have the option to process anything and everything it encounters.
* Additionally, from my experience, some dev tooling can auto-exclude certain directories, such as large git repos, from Windows indexing and reset the exclusion if removed (either that, or the indexer just _nopes_ when dealing with massive directories).  An indexing system whose scope can be externally tampered with, or overrides/ignores the user's intended locations, is problematic especially for power users who intend to use the tool to search for files in specific locations.

---

Sorry for the wall of text; I'd rather dump whatever thoughts I still remember rather than risk forgetting them entirely.

LoZeno commented:
I'm copying and pasting a comment I made on issue #44 , as this is the kind of power tool I really look forward to:

> One of the features that I like from apps like Wox and Ueli, and is missing from the Start menu (or even from Win+S) is that, when using multiple monitors, they open on the monitor where the mouse cursor is present instead of opening on the main monitor.
> This is especially useful when I plug my laptop on a dock, and my eyes are focused on the larger external monitors rather than the "main" smaller laptop monitor. I know I can set any monitor to be "main", but in work locations where hot-desking is the norm it's not very convenient to setup the monitor layout in the Display Settings every time I set my laptop for the day.
> Generally speaking, though, a user's eyes is focused on the screen they were working on last, not always the main screen.
> 
> (That, and the result list of the Windows search changes after being already populated once: I press Win+S, type "Setting", and watch the list of results show up with three results, only to change to 5 results -in a different order, also- after 2 seconds when the indexer completes its job)

crutkas commented:
@LoZeno that is a must have in the spec, it is 21

21 | For multiple monitors, the interface should follow where the mouse cursor is located | 1
-- | -- | --

jyuwono commented:
> How about checking and including 'ueli' (https://github.com/oliverschwendener/ueli) in the comparison matrix as well? 😊

I've tried Ueli and added it to the comparison table. Is there any feature that you really like from Ueli?

jyuwono commented:
> Another app to compare (from back in the day 👴) [AppRocket](https://www.candylabs.com/approcket) 🚀

@ghosttie Since AppRocket is replaced by Skylight, I tried Skylight instead and it has a different experience than other launchers that I've explored. I'm curious, anything that stood out to you in AppRocket/Skylight that should be added in the launcher?

jyuwono commented:
> I would also suggest to take some lessons from [the beta version of Listary](https://discussion.listary.com/t/listary-6-beta/4615).
> 
> Some features I really like is that you can quick select a result by using a Ctrl+Number shortcut:
> 
> Also having to press Shitf+F10 to open the context menu doesn't seem to intuitive for me, a better solution would be to open it by just pressing the Right arrow key. Also Listary (beta) does the context incredibly well as it loads all the options that selecting a file in Explorer would have:

Thanks for the feedback, @JosephTico! Love the idea of Ctrl+Number shortcut. I've added it in the spec (user experience requirement no. 27 and keyboard shortcut requirement no. 9). I've also added your suggestion to have options in context menu to mimic file explorer (user experience requirement no. 37).

ghosttie commented:
@jyuwono I actually much preferred AppRocket to Skylight - AppRocket just launched apps whereas Skylight was a much more ambitious project to let you do a lot more through that kind of interface, and I didn't like that it was slower in exchange for features I didn't really want.

So I guess my most important features from AppRocket would be

1. speed
2. clever algorithm (frecency?) to figure out my most likely match based on only a few keystrokes
3. it included the first letter of words in a shortcut rather than just a "contains" search, so even if the shortcut said "Internet Explorer", searching for "ie" would find it

jyuwono commented:
@HikaruSora great suggestions! 😊

> ### Search Mode
> A search-mode feature, similar to what something like Visual Studio Code uses (whatever they call it), might be worth considering. For context, I'm referring to the ability to search by a specific class of _thing_ much like how VSCode's `Ctrl+P` opens up the palette for file search, `Ctrl+Shift+P` for command search, and `Ctrl+Shift+O` for symbol search. Users can change the mode by starting the query with a particular symbol (for example, `Ctrl+P` opens up the entry field empty but starting the search with `>` changes it to command search).
> 
> An example in the scope of PowerLauncher: users looking for files specifically and may not want to be bombarded by apps or commands which have similar names. Similarly, perhaps they're looking only for the results coming from a particular plugin (file search vs image search vs web search vs command search vs ....) and want to reduce noise in the list.

I've added this suggestion to the spec (user experience requirement no. 29).

> ### Copy as Path
> A common workflow my colleagues and I do is to open explorer, find the target file, `Shift+Right Click` it, and `Copy as path` for pasting into whatever command we have in an open PowerShell or CMD window.
> 
> Having the corresponding option in PowerLauncher's context menu would pair nicely with the file search functionality: `Fuzzy Find File -> Context Menu -> Copy as path` rather than `Fuzzy Find File -> Context Menu -> Open File Location (?) -> Shift+Right Click -> Copy as Path`
> 
> Additionally, having a keyboard shortcut like `Ctrl+Shift+C` to directly copy the focused result's file path to the clipboard would be a plus

We can integrate this feature in V1 and add it as a part of the context menu as well. I've added this as priority 1 in the spec (user experience requirement no. 21-22 and keyboard shortcut requirement no. 3).

> ### Preview
> I saw Niels's Fluent Design mockup of the launcher and it looks fantastic. However, looking at it and comparing to WoX and the other 3rd party launchers, I came to realize that one of the big things that they tend to lack is some sort of preview pane like macOS's Spotlight or another launcher, [Cerebro](https://github.com/KELiON/cerebro). Having a preview makes it easier to check if a particular result indeed contains the information and provides some additional context that can't fit in a single result entry.

This is a good idea but we will need to do more investigation to make sure that performance to show search result is not affected. I've added it to the user experience requirement (no. 38).

> ### Result Density
> From requirement 13,
> 
> > Users can see the icon of the application/file/folder next to the application/file/folder name
> 
> there's no mention about the actual icon size. The mockup and alternate launchers imply that it would be as tall as the result entry and I can't help but think that there should be the option for... the opposite, actually. The nice big results with their icons look really pretty but reduces the number of items that can be displayed at a given time. There's mention of changing text size in Settings requirement 9 but nothing about the icon nor collapsing a result into a single line.
> 
> It seems appropriate to give users the option to increase density, perhaps to something around the size of Spotlight's results, to tilt it towards the Function side of the Function vs Form tradeoff. These _are_ _**power**_ users, after all. A dense, small/no-icon, no-nonsense view mode would match some of them better. It may not fit well with the Fluent Design concept but, well, you might be tossing them directly into PowerShell anyways. 😛

We're planning to give users the ability to customize font colour, font size, and general theme of the launcher in future versions. We can add customizing icon to the list as well. I've added this to settings requirement (no. 12).

jyuwono commented:
> When the system has multiple keyboard layouts installed, map the entered text to the keys pressed, and search for the same key presses across keyboards.
> 
> For example, when the standard Hebrew keyboard is active, pressing the keys code results in the entered text `בםגק`. This could be mapped back for the Hebrew keyboard to `code`, and both terms (`בםגק` and `code`) would be searched for.

@zspitz This is a fantastic suggestion. I've added it to the user experience requirement (no. 28). Our dev team will investigate how this can be done.

keichinger commented:
From @PastorGL
> There is a list at [support.microsoft.com/en-us/help/12445/windows-keyboard-shortcuts](https://support.microsoft.com/en-us/help/12445/windows-keyboard-shortcuts) — looks like almost all Winkey shortcuts are assigned already. Win+\ (backward slash) is the only unused chord on the main keyboard area.

As a German user whenever I see an app using the hotkey `Win+\` I have to remap it to a different key as this key simply doesn't work with a German keyboard layout. Whenever we want to write a `\` we have to press `CTRL+ALT+ß` (that key shares the `?` and `\` characters) - see https://upload.wikimedia.org/wikipedia/commons/thumb/3/36/KB_Germany.svg/1920px-KB_Germany.svg.png for reference in the top right corner (one key to the right of the `0`).

In my opinion it's better to spend some time researching in trying to find a localized hotkey (in case `WIN+Space` doesn't work - I would totally love that hotkey as a work macOS user myself) that works everywhere, that is easy to remember and fast to use, ideally with only two keys (just like `WIN+.`).

ismaelestalayo commented:
> > How about checking and including 'ueli' (https://github.com/oliverschwendener/ueli) in the comparison matrix as well? 😊
> 
> @Kaedin I've tried Ueli and added it to the comparison table. Is there any feature that you really like from Ueli?

I tried Ueli a few months back, and there were two really nice features that would love to see in the PowerLauncher:

- Staring a query with ">" put you in like a "Terminal mode", where whatever you put would be ran on a integrated terminal below the search box. (In this case it could open a new window of the new Windows Terminal with the command from the search box)

- You had the ability to define custom commands to search directly on Youtube, Wikipedia, StackOverflow... so doing a query for "yt kittens" would directly open that query on Youtube in your main browser.

crutkas commented:
This spec is now in our wiki.  https://github.com/microsoft/PowerToys/wiki/Launcher

BillR5GitHub commented:
Note, in the community poll sometimes cited as the reason to close threads objecting to Alt+Space as the default shortcut for Run , Alt+Space received 2 of 16 votes (tied for 3rd) #5331  (although I don't think the poll was well-structured).
Result: [https://www.strawpoll.me/20785284/r](url)
